include/frame.h

```diff
@@ -141,11 +151,163 @@ struct Unit {
 
+/*
+    Terran_Infantry_Armor = 0,
+    Terran_Vehicle_Plating = 1,
+    Terran_Ship_Plating = 2,
+    Terran_Infantry_Weapons = 7,
+    Terran_Vehicle_Weapons = 8,
+    Terran_Ship_Weapons = 9,
+    Zerg_Carapace = 3,
+    Zerg_Flyer_Carapace = 4,
+    Protoss_Ground_Armor = 5,
+    Protoss_Air_Armor = 6,
+    Zerg_Melee_Attacks = 10,
+    Zerg_Missile_Attacks = 11,
+    Zerg_Flyer_Attacks = 12,
+    Protoss_Ground_Weapons = 13,
+    Protoss_Air_Weapons = 14,
+    Protoss_Plasma_Shields = 15,
+*/
+
 struct Resources {
   int32_t ore;
   int32_t gas;
   int32_t used_psi;
   int32_t total_psi;
+  uint64_t upgrades;
+  uint64_t upgrades_level;
+  uint64_t techs;
+
+  // clang-format off
+  enum Upgrades : uint64_t {
+    Terran_Infantry_Armor   = 1ll << 0,
+    Terran_Vehicle_Plating  = 1ll << 1,
+    Terran_Ship_Plating     = 1ll << 2,
+    Zerg_Carapace           = 1ll << 3,
+    Zerg_Flyer_Carapace     = 1ll << 4,
+    Protoss_Ground_Armor    = 1ll << 5,
+    Protoss_Air_Armor       = 1ll << 6,
+    Terran_Infantry_Weapons = 1ll << 7,
+    Terran_Vehicle_Weapons  = 1ll << 8,
+    Terran_Ship_Weapons     = 1ll << 9,
+    Zerg_Melee_Attacks      = 1ll << 10,
+    Zerg_Missile_Attacks    = 1ll << 11,
+    Zerg_Flyer_Attacks      = 1ll << 12,
+    Protoss_Ground_Weapons  = 1ll << 13,
+    Protoss_Air_Weapons     = 1ll << 14,
+    Protoss_Plasma_Shields  = 1ll << 15,
+    U_238_Shells            = 1ll << 16,
+    Ion_Thrusters           = 1ll << 17,
+    Titan_Reactor           = 1ll << 19,
+    Ocular_Implants         = 1ll << 20,
+    Moebius_Reactor         = 1ll << 21,
+    Apollo_Reactor          = 1ll << 22,
+    Colossus_Reactor        = 1ll << 23,
+    Ventral_Sacs            = 1ll << 24,
+    Antennae                = 1ll << 25,
+    Pneumatized_Carapace    = 1ll << 26,
+    Metabolic_Boost         = 1ll << 27,
+    Adrenal_Glands          = 1ll << 28,
+    Muscular_Augments       = 1ll << 29,
+    Grooved_Spines          = 1ll << 30,
+    Gamete_Meiosis          = 1ll << 31,
+    Metasynaptic_Node       = 1ll << 32,
+    Singularity_Charge      = 1ll << 33,
+    Leg_Enhancements        = 1ll << 34,
+    Scarab_Damage           = 1ll << 35,
+    Reaver_Capacity         = 1ll << 36,
+    Gravitic_Drive          = 1ll << 37,
+    Sensor_Array            = 1ll << 38,
+    Gravitic_Boosters       = 1ll << 39,
+    Khaydarin_Amulet        = 1ll << 40,
+    Apial_Sensors           = 1ll << 41,
+    Gravitic_Thrusters      = 1ll << 42,
+    Carrier_Capacity        = 1ll << 43,
+    Khaydarin_Core          = 1ll << 44,
+    Argus_Jewel             = 1ll << 47,
+    Argus_Talisman          = 1ll << 49,
+    Caduceus_Reactor        = 1ll << 51,
+    Chitinous_Plating       = 1ll << 52,
+    Anabolic_Synthesis      = 1ll << 53,
+    Charon_Boosters         = 1ll << 54,
+    Upgrade_60              = 1ll << 60,
+    Unknow                  = 1ll << 62,
+  };
+  enum UpgradesLevel : uint64_t { // could be uint32_t
```

Comment:
I meant move them into UpgradesLevel, it makes more sense to me this way to have all the level upgrades in a single enum.